MIDAS/src/finite-state-machines/fsm.cpp

```diff
@@ -0,0 +1,369 @@
+#include <cmath>
+
+#include "fsm.h"
+#include "thresholds.h"
+#include "systems.h"
+
+
+// helper functions
+
+// most of these helper functions look to average the acceleration values inside the buffer
+
+float FSM::getAcceleration(std::array<HighGData, 8> & hg)
+{ 
+    float avg = 0;
+
+    for (size_t i = 0; i < hg.size(); i++)
+    {
+        avg += (hg.at(i).gz / (float)hg.size());
+    }
+
+    return avg;
+}
+
+float FSM::getAltitude(std::array<Barometer, 8> & bar)
+{
+    float avg = 0;
+
+    for (size_t i = 0; i < bar.size(); i++)
+    {
+        avg += (bar.at(i).altitude / (float)bar.size());
+    }
+
+    return avg;
+}
+
+double FSM::getJerk(std::array<HighGData, 8> & hg)
+{
+
+    double avg1 = 0;
+    for (size_t i = 0; i < hg.size() / 2; i++)
+    {
+        avg1 += (hg.at(i).gz / hg.size());
+    }
+
+    double avg2 = 0;
+    for (size_t i = hg.size() / 2; i < hg.size(); i++)
+    {
+        avg2 += (hg.at(i).gz / hg.size());
+    }
+
+    double sec = pdTICKS_TO_MS(hg.size()) * 1000;
+
+    return (avg2 - avg1) / (sec / 2);
+}
+
+double FSM::getVerticalSpeed(std::array<Barometer, 8> & bar)
+{
+    double avg1 = 0;
+    for (size_t i = 0; i < bar.size() / 2; i++)
+    {
+        avg1 += (bar.at(i).altitude / bar.size());
+    }
+
+    double avg2 = 0;
+    for (size_t i = bar.size() / 2; i < bar.size(); i++)
+    {
+        avg2 += (bar.at(i).altitude / bar.size());
+    }
+
+    double sec = pdTICKS_TO_MS((unsigned int)bar.size()) * 1000;
+
+    return (avg2 - avg1) / (sec / 2);
+}
+
+
+// code for the fsm update function
+
+FSMState FSM::tick_fsm_sustainer(FSMState & state, std::array<HighGData, 8> & hg, std::array<Barometer, 8> & bar){
+    
+    //get current time
+    double current_time = pdTICKS_TO_MS(xTaskGetTickCount());
+    
+    
+    switch (state.curr_state)
+    {
+        case STATE_IDLE:
+            // once a significant amount of acceleartion is detected change states
+            if(getAcceleration(hg) > idle_to_first_boost_acceleration_threshold) {
+                launch_time = current_time;
+                state.curr_state = STATE_FIRST_BOOST;
+            }
+            
+            break;
+
+        case STATE_FIRST_BOOST:
+            acceleration = getAcceleration(hg); 
+
+            // if acceleration spike was too brief then go back to idle
+            if(acceleration < idle_to_first_boost_acceleration_threshold && current_time - launch_time < idle_to_first_boost_time_threshold) {
```

Comment:
Use parenthesis to disambiguate the order of operations.

Reply:
this still